acme/acme.go

```diff
@@ -328,29 +355,44 @@ func (a *ACME) CreateLocalConfig(tlsConfig *tls.Config, checkOnDemandDomain func
 func (a *ACME) getCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
 	domain := types.CanonicalDomain(clientHello.ServerName)
 	account := a.store.Get().(*Account)
-	//use regex to test for wildcard certs that might have been added into TLSConfig
-	for k := range a.TLSConfig.NameToCertificate {
-		selector := "^" + strings.Replace(k, "*.", "[^\\.]*\\.?", -1) + "$"
-		match, _ := regexp.MatchString(selector, domain)
-		if match {
-			return a.TLSConfig.NameToCertificate[k], nil
-		}
-	}
+
+	// Check for existing challenge cert
+	log.Debugf("Checking ACME challenge for %s", domain)
 	if challengeCert, ok := a.challengeProvider.getCertificate(domain); ok {
-		log.Debugf("ACME got challenge %s", domain)
+		log.Debugf("Returning ACME challenge cert for %s", domain)
 		return challengeCert, nil
 	}
+
+	// Check for existing ACME domain cert
+	log.Debugf("Checking ACME domain cert for %v", domain)
 	if domainCert, ok := account.DomainsCertificate.getCertificateForDomain(domain); ok {
-		log.Debugf("ACME got domain cert %s", domain)
+		log.Debugf("Returning ACME domain cert for %s", domain)
 		return domainCert.tlsCert, nil
 	}
+
 	if a.OnDemand {
 		if a.checkOnDemandDomain != nil && !a.checkOnDemandDomain(domain) {
 			return nil, nil
 		}
-		return a.loadCertificateOnDemand(clientHello)
+		cert, err := a.loadCertificateOnDemand(clientHello)
+		if cert != nil && err == nil {
+			return cert, nil
+		}
+		log.Errorf("On demand certicate retrival failed for %v due to %v", domain, err)
 	}
-	log.Debugf("ACME got nothing %s", domain)
+
+	// Check if we have a wildcard cert into TLSConfig that could be used
+	log.Debugf("Checking wildcard certificate matching for %v", domain)
+	for k := range a.TLSConfig.NameToCertificate {
```

Comment:
@CyrilPeponnet, the inversion of certificates checkings you want do has a real interest for the use cases described into the PR but it seems to break the backward compatibily.

If it becomes the default behavior, all users will have to specify a `IgnoreFilters` value.
That's why it may be good to invers the behaviour **ONLY IF** `IgnoreFilters` is specified, so users who will not fill the field will keep the current behavior.

In the way to do this, the wildcard certificate checking could be done in a new function, called at the beginning or the end of `getCertificate` in function of the size of the `IgnoreFilters` list.

The code should look like this :

```go
func (a *ACME) getWildCardCertificate(domain string) (*tls.Certificate, nool) {
	// Check if we have a wildcard cert into TLSConfig that could be used
	log.Debugf("Checking wildcard certificate matching for %v", domain)
	for k := range a.TLSConfig.NameToCertificate {
		selector := "^" + strings.Replace(k, "*.", "[^\\.]*\\.?", -1) + "$"
		match, _ := regexp.MatchString(selector, domain)
		if match {
			log.Debugf("Found a wildcard certificate to use as fallback: %v", k)
			return a.TLSConfig.NameToCertificate[k], true
		}
	}
	return nil, false
}

func (a *ACME) getCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
	domain := types.CanonicalDomain(clientHello.ServerName)
	account := a.store.Get().(*Account)

    if a.IgnoreFilters == nil || a.IgnoreFilters.len == 0 {
		if wildcardCert, ok := a.getWildCardCertificate(domain); ok {
			return wildcardCert, nil
		}
	}
    
	// Check for existing challenge cert
	log.Debugf("Checking ACME challenge for %s", domain)
	if challengeCert, ok := a.challengeProvider.getCertificate(domain); ok {
		log.Debugf("Returning ACME challenge cert for %s", domain)
		return challengeCert, nil
	}

	// Check for existing ACME domain cert
	log.Debugf("Checking ACME domain cert for %v", domain)
	if domainCert, ok := account.DomainsCertificate.getCertificateForDomain(domain); ok {
		log.Debugf("Returning ACME domain cert for %s", domain)
		return domainCert.tlsCert, nil
	}

	if a.OnDemand {
		if a.checkOnDemandDomain != nil && !a.checkOnDemandDomain(domain) {
			return nil, nil
		}
		cert, err := a.loadCertificateOnDemand(clientHello)
		if cert != nil && err == nil {
			return cert, nil
		} else if err != nil {
			log.Errorf("On demand certificate retrieval failed for %v due to %v", domain, err)
		}
	}


    if a.IgnoreFilters != nil && a.IgnoreFilters.len > 0 {
		if wildcardCert, ok := a.getWildCardCertificate(domain); ok {
			return wildcardCert, nil
		}
	}
	log.Debugf("No certificate found to return for %s", domain)
	return nil, nil
}

```
What is your mind about this suggestion?

cc @dtomcej